### DISCUSSION TITLE
Slow Project Editing in Kimai

### DISCUSSION THREAD

**Discussion by markus7777777** [kimai/kimai#4872] [General]
### Describe the issue

Hello everyone!
For the past few weeks, I've been experiencing an issue with Kimai on my local Ubuntu server running in a virtual machine with PHP 8.1. Whenever I select a project within the project list and then click "Edit," it takes a very long time for the editing window to open. During this waiting period, the CPU usage is high for PID php-fpm8.1 , as observed using the _htop_ tool. Interestingly, this issue does not occur when performing similar actions in the customer list or at any other database query.

**Steps to Reproduce:**
- Navigate to the project list.
- Select a project.
- Click on "Edit."
- Observe the loading time and CPU usage.

**Expected Behavior:**
- The editing window should open immediately without high CPU usage.

**Actual Behavior:**
- It takes a long time for the editing window to open, and the CPU is heavily utilized during this time.

**Additional Information:**
- The issue has been occurring for about two weeks.
- The file /var/log/prod.log is empty
- The problem only occurs in the project list, not in the client list

### I already tried

- [X] I've read and searched [the documentation](https://www.kimai.org/documentation/).
- [X] I've searched for similar issues in this repository.
- [X] I've searched for similar issues in [the discussions](https://github.com/kimai/kimai/discussions).

### Kimai version

2.17.0

### How do you run Kimai?

Virtual Server or alike

### Which PHP version are you using?

8.1

**Comment by kevinpapst** [comment]
And what did you do two weeks ago?
Kimai Update? 
Docker Update?
Plugin installation?
New project with Budget?

**Comment by markus7777777** [comment]
I identified the issue with my custom field that uses a dropdown list containing 7,900 possible choices. By reducing the number of dropdown options, the server performance has improved, and it no longer lags as much. Interestingly, there hadn't been any problems with editing or creating projects in Kimai over the past few months. I'm not sure if the issue is related to a recent Kimai update, as I don't use the dropdown list daily, so I can't confirm if the update is the cause.

Example of dropdown options:
01002 I Alsergrund, 01004 I Innere Stadt, 01005 I Josefstadt, 01006 I Landstraße, 01008 I Margarethen, ... (options are separated by commas)

Now, the problem is that after reducing the standard text, I need to reassign all dropdown values for every project.

**Reply by kevinpapst** [reply]
Wow, okay, I never heard that someone has such a huge dropdown list. 
Yeah, the javascript that renders this list is probably getting slow.

> Now, the problem is that after reducing the standard text, I need to reassign all dropdown values for every project.

I don't understand that one.

**Reply by markus7777777** [reply]
I shortened the dropdown list text from **_"01002 I Alsergrund | Wien | ... | ... |"_** to **_"01002 | Alsergrund"_**. This reduction decreased the number of characters from approximately 700k to 200k, resulting in faster loading times for the project editing mask. However, when editing an existing project, I now need to reassign the custom field with the new text because it was shortened.

Assigning more CPU and RAM to the VM did not significantly improve the loading time before I shortened the dropdown list text. Do you have any other suggestions to further speed up the loading time?

Thanks in advance!

**Reply by derStephan** [reply]
>Assigning more CPU and RAM to the VM did not significantly improve the loading time before I shortened the dropdown list text

This is to be expected. javascript is run in one thread on one CPU core only. Adding more cores won't help at all. Faster CPU might but this won't solve your problem on the long run. 

>Do you have any other suggestions to further speed up the loading time?

Reduce the number of items in that field. 7900 Items in a drop-down are rediculous - even in plain HTML which is not in use here. 

You could load parts of that list by ajax to reduce load times.

**Reply by kevinpapst** [reply]
That's the reason why Kimai switches to a AJAX dropdown when using more than 500 tags. 
It just gets too slow. Impossible to fix with a pre-loaded list.

**Reply by markus7777777** [reply]
Thanks for feedback. As expected, reducing the amount of choices improved the performance. A small lag is still existent, but its way faster than before.
